### PR TITLE
docs - discuss the global deadlock detector

### PR DESCRIPTION
**Pull request by lisakowen**
in this PR:
- added fastpath column to pg_locks
- added gp_global_deadlock_detector_period GUC
- added "Global Deadlock Detector" section in Admin Guide "Inserting, Updating, and Deleting Data" topic

questions:
- not sure if i characterized the locking correctly
- the docs do not currently discuss local deadlock.  it isn't clear to me if this has changed, or if it should be documented as part of this effort.
- does GDD output or log any messages when it collects/analyzes lock waiting info?

doc review site links:
- http://docs-lisa-gdd.cfapps.io/600/admin_guide/dml.html#topic_gdd
- http://docs-lisa-gdd.cfapps.io/600/ref_guide/config_params/guc-list.html#gp_global_deadlock_detector_period
- http://docs-lisa-gdd.cfapps.io/600/ref_guide/system_catalogs/pg_locks.html

### CODE REVIEW

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 1]
```diff
@@ -320,6 +321,47 @@
...
     </topic>
   </topic>
+  <topic id="topic_gdd" xml:lang="en">
+    <title>Global Deadlock Detector</title>
+    <body>
+      <p>Detecting deadlock in a distributed system is a complex matter. Previous versions of Greenplum
+        Database prevented global deadlock by holding exclusive table locks for <codeph>UPDATE</codeph>
+        and <codeph>DELETE</codeph> operations. While this strategy does prevent deadlocks, it comes at
+        the cost of poor performance on concurrent updates.</p>
+      <p>Greenplum Database 6.0 includes a global deadlock detector. The Global Deadlock Detector backend
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/dml.xml]: Remove first sentence here, or just remove the version designation.  In general we shouldn't need to qualify functionality in the main docs as being version-specific.

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 2]
```diff
@@ -320,6 +321,47 @@
       </body>
     </topic>
   </topic>
+  <topic id="topic_gdd" xml:lang="en">
+    <title>Global Deadlock Detector</title>
+    <body>
+      <p>Detecting deadlock in a distributed system is a complex matter. Previous versions of Greenplum
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/dml.xml]: I think we can lose the first sentence here.  The rest of this paragraph is probably be better placed in the Release Notes to cover the change in behavior.

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 3]
```diff
@@ -320,6 +321,47 @@
...
+        the cost of poor performance on concurrent updates.</p>
+      <p>Greenplum Database 6.0 includes a global deadlock detector. The Global Deadlock Detector backend
+        process collects lock information on all segments and uses a directed algorithm to detect the
+        existence of global deadlocks. If the Global Deadlock Detector determines that deadlock exists,
+        it breaks the deadlock by cancelling one or more backend processes associated with the youngest 
+        transaction(s) involved.</p>
+      <note>The Greenplum Database 6.0 Global Deadlock Detector removes the concurrent update restriction
+        on heap tables. The Global Deadlock Detector does not remove the concurrent
+        <codeph>UPDATE</codeph>, <codeph>DELETE</codeph>, and <codeph>SELECT ... FOR UPDATE</codeph>
+        restriction on AO/CO tables; Greenplum Database base still employs table-level locking on these
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/dml.xml]: Stray word here: "base"

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 4]
```diff
@@ -320,6 +321,47 @@
...
+        operations.</note>
+      <p>The Global Deadlock Detector backend process is automatically started on the master host when
+        you start Greenplum Database. You configure the interval at which the Global Deadlock Detector
+        collects and analyzes lock waiting data via the
+         <codeph><xref href="../ref_guide/config_params/guc-list.xml#gp_global_deadlock_detector_period" type="section">gp_global_deadlock_detector_period</xref></codeph>
+         server configuration parameter.</p>
+      <p>To view lock waiting information for all segments, run the <codeph>pg_dist_wait_status()</codeph>
+        user-defined function. You can use the output of this function to determine which transactions
+        are waiting on locks, which transactions are holding locks, and the segments on which the
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/dml.xml]: and the segments on which the transactions are executing -> and which segments are executing the transactions

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 5]
```diff
@@ -320,6 +321,47 @@
...
+      <p>Detecting deadlock in a distributed system is a complex matter. Previous versions of Greenplum
+        Database prevented global deadlock by holding exclusive table locks for <codeph>UPDATE</codeph>
+        and <codeph>DELETE</codeph> operations. While this strategy does prevent deadlocks, it comes at
+        the cost of poor performance on concurrent updates.</p>
+      <p>Greenplum Database 6.0 includes a global deadlock detector. The Global Deadlock Detector backend
+        process collects lock information on all segments and uses a directed algorithm to detect the
+        existence of global deadlocks. If the Global Deadlock Detector determines that deadlock exists,
+        it breaks the deadlock by cancelling one or more backend processes associated with the youngest 
+        transaction(s) involved.</p>
+      <note>The Greenplum Database 6.0 Global Deadlock Detector removes the concurrent update restriction
```
Comment (dyozie) [inline gpdb-doc/dita/admin_guide/dml.xml]: This note needs some re-wording. It seems rather that the Deadlock Detector is only used for heap tables, while full table locking is used for AO/CO tables and includes restrictions to concurrent updates?  If that's the case, better to describe the detector as a heap-table only feature from the start.

**Comment by pivotal-ning-yu** [comment]
I'm reviewing this doc, please give me some time.

**Comment by lisakowen** [comment]
thank you for reviewing @pivotal-ning-yu .  i reorg'd the content a little, and added some info about local deadlock detection.  if you get a chance, can you do a quick review of the first couple of paragraphs?

File: gpdb-doc/dita/admin_guide/dml.xml [hunk 6]
```diff
@@ -320,6 +321,46 @@
...
     </topic>
   </topic>
+  <topic id="topic_gdd" xml:lang="en">
+    <title>Global Deadlock Detector</title>
+    <body>
+      <p>The Greenplum Database Global Deadlock Detector backend
+        process collects lock information on all segments and uses a directed algorithm to detect the
+        existence of local and global deadlocks. This algorithm allows Greenplum Dabase to relax
+        concurrent update restrictions on heap tables. (Greenplum Database still employs table-level
+       locking on AO/CO tables, restricting concurrent <codeph>UPDATE</codeph>,
```
Comment (gaos1) [inline gpdb-doc/dita/admin_guide/dml.xml]: most likely we need table lock for 4 cases. two of them are listed here. another two cases are under discussion in this thread. we will update you later.  https://groups.google.com/a/greenplum.org/forum/#!topic/gpdb-dev/LnkABMhIHcY

**Comment by pivotal-ning-yu** [comment]
On Thu, Dec 27, 2018 at 3:39 AM Lisa Owen <notifications@github.com> wrote:

> thank you for reviewing @pivotal-ning-yu
> <https://github.com/pivotal-ning-yu> . i reorg'd the content a little,
> and added some info about local deadlock detection. if you get a chance,
> can you do a quick review of the first couple of paragraphs?
>
Thanks for the updates, looks perfect, +1.

**Pull request merged**